feat:[D0C-3969]: Adding when condition for looping strategy (#5394)
harness/developer-hub: docs/platform/pipelines/looping-strategies/when-conditions-inside-looping-strategies.md

---
title: Using when conditions inside looping strategy
description: Run conditional execution inside looping strategy
sidebar_position: 3
---

> You can use conditional execution inside [looping strategy](/docs/platform/pipelines/looping-strategies/looping-strategies-matrix-repeat-and-parallelism.md) by providing `when` keyword inside your strategy. 

Reviewer: I think we need more context here. When do you use the When inside the Looping Strategy text box, and when do you use Conditional Execution? The order is a bit counter intuitive here, so we need to explain.

---

> title: Using when conditions inside looping strategy

Author: sure, thanks for the info @brian-f-harness . I will keep in mind from next PR.

    - stage:
        name: sA
        identifier: sA
        description: ""
        type: Custom
        spec:
          execution:
            steps:
              - step:
                  type: ShellScript
                  name: ShellScript_1
                  identifier: ShellScript_1
                  spec:
                    shell: Bash
                    executionTarget: {}
                    source:
                      type: Inline
                      spec:
                        script: echo hey
                    environmentVariables: []
                    outputVariables: []
                  timeout: 10m
        tags: {}
        strategy:
          when: <+pipeline.name> == "looping_strategy_when"
          matrix:
            jdk:
              - "18"
              - "17"
              - "16"
              - "15"
              - "14"
              - "13"
              - "12"
              - "11"
              - "10"
              - "9"
        when:
          pipelineStatus: Success
          condition: <+strategy.iteration> == 2
  allowStageExecutions: true
```
Let's discuss how this YAML is going to get evaluated and what will be the expected output:-

> 1. First, it will check whether ``when: <+pipeline.name> == "looping_strategy_when"`` is true, if it's true then only it will move forward to run the looping strategy else it will get skipped. 

Reviewer 2: ```suggestion
1. First, it will check whether ``when: <+pipeline.name> == "looping_strategy_when"`` is true. If it's true, then it will move forward to run the looping strategy, otherwise it will be skipped. 
```

![](./static/looping_stretegy_when_conditions.png)

> 2. Second, if we have a stage level conditional execution as provided in the above example:-

Reviewer 2: ```suggestion
2. Second, if you have a stage-level conditional execution as provided in the above example:
```

```yaml
 when:
        pipelineStatus: Success
        condition: <+strategy.iteration> == 2
```

> Here, after the first ``when: <<+pipeline.name> == "looping_strategy_when"`` condition is evaluated then only it will evaluate the stage level condition.

Reviewer 2: ```suggestion
Here, after the first ``when: <<+pipeline.name> == "looping_strategy_when"`` condition is evaluated, then it will only evaluate the stage-level condition.
```

> To summarize, the above yaml will first check if the pipeline name matches ``looping_strategy_when`` and if it does i.e if the status is ``Success`` upto this point it will run the 2nd element of the matrix. 

Reviewer 2: ```suggestion
To summarize, the above YAML will first check if the pipeline name matches ``looping_strategy_when``, and if it does, for example, if the status is ``Success``, it will run the second element of the matrix. 
```

![](./static/looping_strategy_when.png)

> 2. Let's consider an example where it will skip to run looping strategy:-

Reviewer 2: ```suggestion
Let's consider an example where it will skip to run looping strategy:
```

  identifier: looping_strategy_when
  tags: {}
  projectIdentifier: fdsf
  orgIdentifier: default
  stages:
    - stage:
        name: sA
        identifier: sA
        description: ""
        type: Custom
        spec:
          execution:
            steps:
              - step:
                  type: ShellScript
                  name: ShellScript_1
                  identifier: ShellScript_1
                  spec:
                    shell: Bash
                    executionTarget: {}
                    source:
                      type: Inline
                      spec:
                        script: echo hey
                    environmentVariables: []
                    outputVariables: []
                  timeout: 10m
        tags: {}
        strategy:
          when: <+pipeline.name> == "feat1"
          matrix:
            jdk:
              - "18"
              - "17"
              - "16"
              - "15"
              - "14"
              - "13"
              - "12"
              - "11"
              - "10"
              - "9"
        when:
          pipelineStatus: Success
          condition: <+strategy.iteration> == 2
  allowStageExecutions: true
```

> In this example, the pipeline name doesn't matches as ``feat1`` therefore it will skip to run the looping strategy in this case ``matrix`` and move forward to run the rest of the pipeline. 

Reviewer 2: ```suggestion
In this example, the pipeline name doesn't match as ``feat1``, therefore it will skip running the looping strategy (in this case ``matrix``), and move forward to run the rest of the pipeline. 
```
